File: Timetable_naming_trend.m
function STN_rd=Timetable_naming_trend(merged_data)

% change the parameter's names from ebas into the naming convention for the
% trend analysis

%INPUT: merged_data=structure with data read from Netcdf file from ebas with all data
%from one station

% OUTPUT: STN_rd: timetable with correct parameter's names

% Exemple: STN_rd=Timetable_naming_trend(merged_data);


% read the wavelength (second dimention
Ni=contains({merged_data.Variables.Name},'Wavelength');
Wavelength=merged_data.Variables(Ni).Value;


%% compute time from ebas format (days from 1900 1 1)
% transform the structure with all wavelengths to a table with the right
% naming convention
for i=1:length(merged_data.Variables)
    if ~strcmp(merged_data.Variables(i).Name,'d_Wavelength')

        if strcmp(merged_data.Variables(i).Name,'time')
            STN_rds.Time=datetime(1900,1,1) + days(merged_data.Variables(i).Value); % compute time as datetime
        else

            for j=1:size(merged_data.Variables(i).Value,2) % take all data from an optical parameter

                % select variable type
                if contains(merged_data.Variables(i).Name,'_scattering')
                    var='Bs';
                elseif contains(merged_data.Variables(i).Name,'backscattering')
                    var='Bbs';
                elseif contains(merged_data.Variables(i).Name,'absorption')
                    var='Ba';
                elseif contains(merged_data.Variables(i).Name,'humidity')
                    var='U';
                else
                    warning('none of the variable has been identified')
                end

                % select size cut
                if contains(merged_data.Variables(i).Name,'pm10')
                    pm='0';
                elseif contains(merged_data.Variables(i).Name,'pm1')
                    pm='1';
                elseif contains(merged_data.Variables(i).Name,'pm2.5')
                    pm='2';
                else
                    warning('size cut is identified as TSP')
                    pm='';
                end

                %select wavelength
                if Wavelength(j)==370
                    wv='1';
                elseif Wavelength(j)==470
                    wv='2';
                elseif Wavelength(j)==520
                    wv='3';
                elseif Wavelength(j)==590
                    wv='4';
                elseif Wavelength(j)==660
                    wv='5';
                elseif Wavelength(j)==880
                    wv='6';
                elseif Wavelength(j)==950
                    wv='7';
                else
                    if Wavelength(j)<500 
                        wv='B';
                    elseif 500<Wavelength(j) && Wavelength(j)<600
                        wv='G';
                    elseif 600<=Wavelength(j) && Wavelength(j)<720
                        wv='R';
                    elseif Wavelength(j)>720
                        wv='Q';
                    end
                
                end

                % keep the paramerter only if there is data for the applied wavelength
                if sum(isnan(merged_data.Variables(i).Value(:,j)),1)< size(merged_data.Variables(i).Value(:,j),1)
                    name_var=strcat(var,wv,pm,'_I'); % parameter name
                    STN_rds.(name_var)=merged_data.Variables(i).Value(:,j);
                end
            end
        end
    end

end

% Convert the table into timetable
STN_rd=table2timetable(struct2table(STN_rds));

File: Timetable_naming_trend.py
import numpy as np
import pandas as pd


def timetable_naming_trend(ds):

    data = {}

    data["Time"] = pd.to_datetime(ds["time"].values)

    for var_name in ds.data_vars:

        da = ds[var_name]
        
        if da.dims == ("d_Wavelength", "time"):

            wavelength = ds["d_Wavelength"].values

            da = da.transpose("time", "d_Wavelength")

            values = da.values
    
            # keep variable wavelengths
            mask = ~np.isnan(values).all(axis=0)
    
            wavelength = wavelength[mask]
            values = values[:, mask]
    
            for j, wl in enumerate(wavelength):
            
                col = values[:, j]
    
                # keep the paramerter only if there is data for the applied wavelength
                if np.isnan(col).all():
                    continue
                
                # select variable type
                if "_scattering" in var_name:
                    prefix = "Bs"
                    inst = "S"
    
                elif "backscattering" in var_name:
                    prefix = "Bbs"
                    inst = "S"
    
                elif "absorption" in var_name:
                    prefix = "Ba"
                    inst = "A"
    
                elif "humidity" in var_name:
                    prefix = "U"
    
                else:
                    continue
                
                # select size cut
                if "pm10" in var_name:
                    pm = "0"
    
                elif "pm2.5" in var_name:
                    pm = "2"
    
                elif "pm1" in var_name:
                    pm = "1"
    
                else:
                    pm = ""
    
                # select wavelength
                wl_map = {370: "1",470: "2",520: "3",
                        590: "4",660: "5",880: "6",950: "7",}
                
                if len(wavelength) in [3, 7, 10]:
                
                    if wl in wl_map:
                        wv = wl_map[wl]
    
                    else:
                        if wl < 500:
                            wv = "B"
                        elif wl < 600:
                            wv = "G"
                        elif wl < 720:
                            wv = "R"
                        else:
                            wv = "Q"
    
                elif set(wl_map).issubset(set(wavelength)):
                    if wl in wl_map:
                        wv = wl_map[wl]
                    else:
                        wv = int(wl)
                else:
                    wv = int(wl)
    
                name = f"{prefix}{wv}{pm}_{inst}"
    
                data[name] = col
            
            # save overlapped 660nm data in a new column
            if prefix =='Ba':
                for i in [0,1,2]:
                    if (f'Ba1{i}_A' in data and f'Ba5{i}_A' in data and
                        np.count_nonzero(~np.isnan(data[f'Ba1{i}_A'])) < 
                        np.count_nonzero(~np.isnan(data[f'Ba5{i}_A']))):
    
                        mask = np.isnan(data[f'Ba1{i}_A'])
    
                        data[f'Ba660{i}_A'] = np.where(mask, data[f'Ba5{i}_A'], np.nan)
                        data[f'Ba5{i}_A'] = np.where(mask, np.nan, data[f'Ba5{i}_A'])

    df = pd.DataFrame(data)

    df = df.set_index("Time")

    return df